### PR TITLE
descs: use catalog lookups when batch hydrating

### PR DESCRIPTION
**Pull request by jasonmchan**
Previously, we hydrated sets of descriptors with
`HydrateGivenDescriptors`, which used a slice of descriptors. This
commit adds `hydrateCatalog`, which uses a catalog to look up
descriptors and hydrates the catalog in place.

Release note: None

### CODE REVIEW

File: pkg/sql/catalog/descs/hydrate.go [hunk 1]
```diff
@@ -232,3 +233,45 @@ func HydrateGivenDescriptors(ctx context.Context, descs []catalog.Descriptor) er
...
+				tree.TypeObject)
+		}
+		dbDesc := cat.LookupDescriptorEntry(typDesc.GetParentID())
+		if dbDesc == nil {
+			n := fmt.Sprintf("[%d]", typDesc.GetParentID())
+			return tree.TypeName{}, nil, sqlerrors.NewUndefinedDatabaseError(n)
+		}
+		scDesc := cat.LookupDescriptorEntry(typDesc.GetParentSchemaID())
+		name := tree.MakeQualifiedTypeName(dbDesc.GetName(), scDesc.GetName(), typDesc.GetName())
+		return name, typDesc.(catalog.TypeDescriptor), nil
```
Comment (ajwerner) [inline pkg/sql/catalog/descs/hydrate.go:257]: Can you check this type assertion and return an error if it is not a `TypeDescriptor`?

**Review by ajwerner — COMMENTED** [review]
May I request some testing? Doesn't have to be a ton. One approach would be to spin up a test server, create some descriptors, pull them down, shove different subsets of them into a catalog for different subtests and make sure the right thing happens.

**Comment by jasonmchan** [comment]
> May I request some testing? Doesn't have to be a ton. One approach would be to spin up a test server, create some descriptors, pull them down, shove different subsets of them into a catalog for different subtests and make sure the right thing happens.

I wrote a test, but not sure if it's what you had in mind. What do you mean by pulling down descriptors?

File: pkg/sql/catalog/descs/hydrate.go (hunk 1 above)
Comment (jasonmchan) [reply pkg/sql/catalog/descs/hydrate.go]: Done.

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84850)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jasonmchan)

___
*[`pkg/sql/catalog/descs/collection_test.go` line 809 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84850#-N7vqK3Y9fGw6kA8f5sj:-N7vqK3Y9fGw6kA8f5sk:b-qb6ag0) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc488f156324f2907983c38be5bf7fd329b4a1e/pkg/sql/catalog/descs/collection_test.go#L809)):*
> ```Go
> 
> 	execCfg := s.ExecutorConfig().(sql.ExecutorConfig)
> 	require.NoError(t, sql.DescsTxn(ctx, &execCfg, func(
> ```

This test gives you some coverage of the positive cases but not of the error cases. I was suggesting:

```
        var cat nstree.Catalog
	require.NoError(t, sql.DescsTxn(ctx, &execCfg, func(
		ctx context.Context, txn *kv.Txn, descriptors *descs.Collection,
	) (err error) {
		// The descriptors in this catalog should be hydrated.
		cat, err = catkv.GetCatalogUnvalidated(ctx, execCfg.Codec, txn)
		return err
       })
```
Now build a table of test cases where you construct new catalogs with either subsets of the read catalog or with some injected badness. Maybe have a function which takes the catalog and mucks with it. The test harness can clone the catalog for each case. In particular, it'd be cool to exercise the case where a descriptor you expect to be a type is not. WDYT?

I imagine one consideration might be that the function to hydrate a catalog is unexported and this is an external test. The workaround is to export it via a `helper_test.go` variable. See https://github.com/cockroachdb/cockroach/blob/19541357fa74af84adbf20561cb5794c611aa7dd/pkg/upgrade/upgrades/helpers_test.go#L32 for an example of this. It's a nice technique and does not export the symbol outside of testing of the package.

**Review by jasonmchan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84850)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[`pkg/sql/catalog/descs/collection_test.go` line 809 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84850#-N7vqK3Y9fGw6kA8f5sj:-N7wVN8M-5oAYyhH0_Vo:b-pf3rw9) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dc488f156324f2907983c38be5bf7fd329b4a1e/pkg/sql/catalog/descs/collection_test.go#L809)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This test gives you some coverage of the positive cases but not of the error cases. I was suggesting:

```
        var cat nstree.Catalog
	require.NoError(t, sql.DescsTxn(ctx, &execCfg, func(
		ctx context.Context, txn *kv.Txn, descriptors *descs.Collection,
	) (err error) {
		// The descriptors in this catalog should be hydrated.
		cat, err = catkv.GetCatalogUnvalidated(ctx, execCfg.Codec, txn)
		return err
       })
```
Now build a table of test cases where you construct new catalogs with either subsets of the read catalog or with some injected badness. Maybe have a function which takes the catalog and mucks with it. The test harness can clone the catalog for each case. In particular, it'd be cool to exercise the case where a descriptor you expect to be a type is not. WDYT?

I imagine one consideration might be that the function to hydrate a catalog is unexported and this is an external test. The workaround is to export it via a `helper_test.go` variable. See https://github.com/cockroachdb/cockroach/blob/19541357fa74af84adbf20561cb5794c611aa7dd/pkg/upgrade/upgrades/helpers_test.go#L32 for an example of this. It's a nice technique and does not export the symbol outside of testing of the package. 
</blockquote></details>

This makes sense! Wrote these tests up.

File: pkg/sql/catalog/descs/collection_test.go [hunk 2]
```diff
@@ -783,3 +788,110 @@ func TestDescriptorCache(t *testing.T) {
...
+	tdb.Exec(t, `CREATE TYPE db.schema.typ AS ENUM ('a', 'b')`)
+	tdb.Exec(t, `CREATE TABLE db.schema.table(x db.schema.typ)`)
+
+	execCfg := s.ExecutorConfig().(sql.ExecutorConfig)
+	t.Run("invalid catalog", func(t *testing.T) {
+		type catalogTamperFn func(nstree.Catalog) nstree.Catalog
+		deleteDescriptor := func(name string) catalogTamperFn {
+			return func(cat nstree.Catalog) nstree.Catalog {
+				var descToDelete catid.DescID
+				cat.ForEachDescriptorEntry(func(desc catalog.Descriptor) error {
```
Comment (ajwerner) [inline pkg/sql/catalog/descs/collection_test.go:817]: The linter wants `_ =` here and below

**Review by ajwerner — APPROVED** [review]
LGTM!

**Pull request merged**